zerver/tests/test_zcommand.py

```diff
@@ -80,3 +83,93 @@ def test_fixed_zcommand(self) -> None:
+        self.subscribe(hamlet, "general")
+        test = self.make_stream("test")
+        self.subscribe(hamlet, "test")
+
+        payload = dict(
+            command="/digress",
+            command_data=orjson.dumps(
+                dict(
+                    old_stream="general",
+                )
+            ).decode(),
+        )
+        result = self.client_post("/json/zcommand", payload)
+        self.assert_json_error(result, "Invalid data.")
+
+        payload = dict(
+            command="/digress",
+            command_data=orjson.dumps(
+                dict(
+                    old_stream="general",
+                    old_topic="old topic name",
+                    new_stream="general",
+                    new_topic="old topic name",
+                )
+            ).decode(),
+        )
+        result = self.client_post("/json/zcommand", payload)
+        self.assert_json_success(result)
+        self.assertIn("Cannot digress to the same thread.", result.json()["msg"])
+
+        payload = dict(
+            command="/digress",
+            command_data=orjson.dumps(
+                dict(
+                    old_stream="general",
+                    old_topic="old topic name",
+                    new_stream="test",
+                    new_topic="new topic name",
+                )
+            ).decode(),
+        )
+        result = self.client_post("/json/zcommand", payload)
+        self.assert_json_success(result)
+        self.assertIn("Old topic #**general>old topic name** does not exist.", result.json()["msg"])
+
+        self.client_post(
+            "/json/messages",
+            {
+                "type": "stream",
+                "to": "general",
+                "client": "test suite",
+                "content": "Test message",
+                "topic": "old topic name",
+            },
+        )
+
+        payload = dict(
+            command="/digress",
+            command_data=orjson.dumps(
+                dict(
+                    old_stream="general",
+                    old_topic="old topic name",
+                    new_stream="test",
+                    new_topic="new topic name",
+                )
+            ).decode(),
+        )
+        result = self.client_post("/json/zcommand", payload)
+        self.assert_json_success(result)
+        self.assertEqual("", result.json()["msg"])
+
+        messages = get_topic_messages(hamlet, general, "old topic name")
+        self.assertEqual(len(messages), 2)
+        self.assertEqual(
+            messages[1].content,
+            f"@_**King Hamlet|{hamlet.id}** will be talking on the new topic: #**test>new topic name**",
+        )
+
+        messages = get_topic_messages(hamlet, test, "new topic name")
+        self.assertEqual(len(messages), 1)
+        self.assertEqual(
+            messages[0].content,
+            f"@_**King Hamlet|{hamlet.id}** digressed this from old topic: #**general>old topic name**",
+        )
```

Comment:
We'll want to expand the tests to have a case where one tries to digress from a message in a private stream without shared history that the user joined after the original message, which should fail.  And then maybe another for "private stream with shared history" where the user doesn't have a UserMessage -- cleanest would be for that to reuse the data setup and just toggle the stream to "shared history" in between attempts.

Reply:
I split this into multiple tests, because the `UserMessage` generated while testing the happy path for the shared history case would interfere while testing the failure condition of the private history case.